### DISCUSSION TITLE
MIDI Thru

### DISCUSSION THREAD

**Discussion by Sedaxel** [rerdavies/pipedal#133] [Q&A]
First, a giant thank you to Robin for creating this awesome project.

I have the PiPedal installed in a Pi4, Android app running in my smartphone, several LV2 plugins installed, etc. I chose Ubuntu server as OS. My audio interface is a Mooer GE250, which also has MIDI.

Now I'm trying to add a midi foot controller (MVAVE Chocolate) which would control PiPedal parameters and presets, but also would change presets in the GE250... So I need MIDI messages coming from the MVAVE to reach not only PiPedal, but also GE250.

Tried with aconnect, but when PiPedal has the MVAVE assigned, this can't be linked with aconnect to the GE250.

Is it possible to get PiPedal to let incoming MIDI messages pass through to other MIDI devices?. That is... Is it possible for PiPedal to have midi thru?

**Comment by stefets** [comment]
You need a Midi router that will configure this behaviour for that. Mididings is the best choice.

**Reply by rerdavies** [reply]
PiPedal runs as a daemon (so that it can run as an IoT device), Getting a midi router set up before PiPedal binds to its ALSA midi port might not be that  straightforward.

**Reply by stefets** [reply]
I will do some tests on my side and I will update this thread with my results.

**Comment by rerdavies** [comment]
Yeah... I think I can do that. 

Let me just confirm that I understand the details....

1)  The MWAVE chocolate is a Bluetooth connected foot controller with 4 stomp switches, which I'm guessing you will program to send midi Program Changes messages for the GE250, and either program changes or toggle controls for PiPedal.

2) You want to select the MWAVE chocolate as input device in PiPedal, and the GE-250 USB channel as a MIDI through device in PiPedal. 

PiPedal will then monitor the MWAVE Chocolate midi channel, looking for midi triggers, but will ALSO forward any messages it receives to the GE250 ALSA channel (which looks like it will only usefully supports MIDI Program Change messages). There will be no particularly guarantees with respect to latency or jitter (tens of milliseconds, not one or two milliseconds), which should be ok because you only have footswitches.

If that's what you want, that seems reasonably doable. 

Before I cut code, could I get you to confirm that the following things work:

1) That PiPedal can successfully read from the MWAVE Chocolate ALSA midi input.

2) That the GE250 will do what you want if you patch  MIDI input from the ALSA MWAVE Chocolate MIDI input to the GE250 ALSA USB MIDI output using standard LINUX utilities. You should be able to do this with QJackctl, Carla, or Ardour, but I'm afraid I will leave it to you to work out the exact details of how to do it.  [edit: or aconnect. That works fine too.].

It's a fairly specific ask,  but not that difficult to do.I would like some assurance that it's going to actually do what you want if I implement it. The G250 user manual seems especially unhelpful with respect to what the "PC Mapping" feature does. And I can think of a dozen ways you can be disappointed by what the GE250 does.  So trying things out with something other than PiPedal  is an important step.

Let me know if you can get it to do what you want it to do, and I shall make it so in PiPedal.

**Reply by Sedaxel** [reply]
Wow... what a quick response from you guys. Thank you.

Rerdavies, yes, you are understanding perfectly. Only, I´m connecting the MVAVE via USB (I avoid the bluetooth for now).

1. I´ve tested the MVAVE correctly works with PiPedal. I select MVAVE as a MIDI input in PiPedal, and PiPedal reads messages perfectly. By the way, PiPedal also sees GE250 (and it works too as a MIDI input if I set the MIDI mode in GE250 to MIDI OUT).

2. If I deselect MVAVE from PiPedal, I can use "aconnect" command to link MVAVE to GE250, and they work well together. I send messages from MVAVE to GE250 and it works.

But, if they both are linked this way, then PiPedal doesn´t allow to select MVAVE as a MIDI input.

Yeah, I know GE250 MIDI capabilities. I can use PC messages to change presets, and I have the CC messages implementation chart for other tasks (effects on/off and whatever). No problem in that part. Moreover, I´m planning to keep it simple and just select presets. That´s going to be the easy part... hehehe

Thank you!
